Sources/NIOCore/EventLoopFuture+AssumeIsolated.swift

```diff
@@ -458,34 +478,70 @@ extension EventLoopFuture {
     }
 
-    /// Assumes the calling context is isolated to the future's event loop.
-    @usableFromInline
-    func assumeIsolated() -> Isolated {
+    /// Returns a variant of this ``EventLoopFuture`` with less strict
+    /// `Sendable` requirements. Can only be called from on the
+    /// ``EventLoop`` to which this ``EventLoopFuture`` is bound, will crash
+    /// if that invariant fails to be met.
+    @inlinable
+    public func assumeIsolated() -> Isolated {
+        self.eventLoop.preconditionInEventLoop()
+        return Isolated(_wrapped: self)
+    }
+
+    /// Returns a variant of this ``EventLoopFuture`` with less strict
+    /// `Sendable` requirements. Can only be called from on the
+    /// ``EventLoop`` to which this ``EventLoopFuture`` is bound, will crash
+    /// if that invariant fails to be met.
+    ///
+    /// This is an unsafe version of ``EventLoopFuture/assumeIsolated()`` which
+    /// omits the runtime check in release builds. This improves performance, but
+    /// should only be used sparingly.
+    @inlinable
+    public func assumeIsolatedUnsafeUnchecked() -> Isolated {
         self.eventLoop.assertInEventLoop()
         return Isolated(_wrapped: self)
     }
 }
 
+@available(*, unavailable)
+extension EventLoopFuture.Isolated: Sendable {}
+
 extension EventLoopPromise {
     /// A struct wrapping an ``EventLoopPromise`` that ensures all calls to any method on this struct
     /// are coming from the event loop of the promise.
-    @usableFromInline
-    struct Isolated {
+    ///
+    /// This type is explicitly not `Sendable`. It may only be constructed on an event loop,
+    /// using ``EventLoopPromise/assumeIsolated``, and may not subsequently be passed to other isolation
+    /// domains.
+    ///
+    /// Using this type relaxes the need to have the promise completion functions accept ``Sendable``
+    /// values, as this type can only be handled on the ``EventLoop``.
+    ///
+    /// This type does not offer the full suite of completion functions that ``EventLoopPromise``
+    /// does, as many of those functions do not require `Sendable` values already. It only offers
+    /// versions for the functions that do require `Sendable` types. If you have an
+    /// ``EventLoopPromise/Isolated`` but need a regular ``EventLoopPromise``, use
+    /// ``EventLoopPromise/Isolated/nonisolated()`` to unwrap the value.
+    public struct Isolated {
         @usableFromInline
         let _wrapped: EventLoopPromise<Value>
 
+        @inlinable
+        init(_wrapped: EventLoopPromise<Value>) {
+            self._wrapped = _wrapped
+        }
+
         /// Deliver a successful result to the associated `EventLoopFuture<Value>` object.
         ///
         /// - Parameters:
         ///   - value: The successful result of the operation.
         @inlinable
-        func succeed(_ value: Value) {
-            self._wrapped.futureResult.eventLoop.assertInEventLoop()
+        public func succeed(_ value: Value) {
```

Comment:
We don't have a `fail(_:)` method here. `Error`s are always `Sendable` so I suppose it's technically not necessary but it still feels like an odd omission given we have `completeWith(_:)`.

Reply:
Yeah, I almost added it and then concluded that it was nicer if we only had methods on this type that dropped the `Sendable` constraint. `completeWith` meets that bar, as it drops the `Sendable` requirement on the type, which _is_ otherwise present.

Reply:
I agree. You can always get the underlying promise back by calling `.nonisolated.fail()`